File: merge_sort.js
/*
Merge Sort
----------
Advantages:
    - Relatively fast for large data sets
    - Stable
Disadvantages:
    - Not in-place
Time complexity:
    - worst:   O(n*log(n))
    - average: O(n*log(n))
    - best:    O(n*log(n))
Space complexity:
    - O(n)
*/

function merge(left, right) {
    var result = [];
    while (left.length > 0 || right.length > 0) {
        if (left.length > 0 && right.length > 0) {
            if (left[0] < right[0]) {
                result.push(left.shift());
            } else {
                result.push(right.shift());
            }
        } else if (left.length > 0) {
            result.push.apply(result, left);
            break;
        } else {
            result.push.apply(result, right);
            break;
        }
    }
    return result;
}

function merge_sort(l) {
    if (l.length <= 1) {
        return l;
    } else {
        var pivot = l.length/2;
        var left = merge_sort(l.slice(0, pivot));
        var right = merge_sort(l.slice(pivot));
        return merge(left, right);
    }
}


//test cases
l = [10, 9, 8, 7, 6, 5, 4, 3, 2, 1];
console.log('Unsorted list: ', l);
l = merge_sort(l);
console.log('Sorted list: ', l);

File: merge_sort.py
def merge(left, right):
    result = [];
    while len(left) > 0 or len(right) > 0:
        if len(left) > 0 and len(right) > 0:
            if left[0] < right[0]:
                result.append(left.pop(0))
            else:
                result.append(right.pop(0))
        elif len(left) > 0:
            result.extend(left)
            break
        else:
            result.extend(right)
            break
    return result

def merge_sort(l):
    """
    Merge Sort
    ----------
    Advantages:
        - Relatively fast for large data sets
        - Stable
    Disadvantages:
        - Not in-place
    Time complexity:
        - worst:   O(n*log(n))
        - average: O(n*log(n))
        - best:    O(n*log(n))
    Space complexity:
        - O(n)
    """
    if len(l) <= 1:
        return l
    else:
        pivot = len(l)//2
        left = merge_sort(l[:pivot])
        right = merge_sort(l[pivot:])
        return merge(left, right)

if __name__ == '__main__':
    # test cases
    l = list(range(10, 0, -1))
    print('Unsorted list: ' + str(l));
    l = merge_sort(l)
    print('Sorted list: ' + str(l))

